🚀 1단계 - TDD로 문자열 덧셈 계산기 구현

rima361 opened the pull request:
### 임금님 귀는 당나귀 귀
지난번 강의 들었을때 Class 에 인격화를 하지 말라고 했던것 같은데, 공감이 아직 안가서 갈팡질팡 하는 부분이 있어요.
예를 들면 `Support vs Supporter` ....

Commit:
feat: String 클래스에 대한 학습 테스트
- 문자열을 ,로 split 했을 때 분리하는 기능, 테스트 구현

Commit:
test: String subString, chatAt 메서드 테스트 구현

Commit:
feat: Set Collection에 대한 학습 테스트
- Set size, contains 메서드 테스트 구현
- ParameterizedTest, CsvSource 활용
- package 이동

Commit:
feat: 문자열 덧셈 계산기 기능, 테스트 구현

Commit:
refactor: 빈문자열 확인 메서드 를 String class 로 대체하여 활용

jayden-17 commented on src/main/java/caculator/PositiveNumber.java:
```diff
@@ -0,0 +1,29 @@
+public class PositiveNumber {
+    private final long value;
+
+    public PositiveNumber(final String value) {
```
wrapper class 👍

jayden-17 commented on src/main/java/caculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,13 @@
+    public static long add(final String value) {
+        return Arrays.stream(StringSplitter.split(value))
+                     .map(PositiveNumber::new)
+                     .map(PositiveNumber::getValue)
```
PositiveNumber 가 sum 을 직접할 수도 있지 않을까요 ? 

이 부분은 같이 고민해보면 좋을 것 같아요 ~ 

wrapper class 에서 getValue 로 기본 값을 노출하는 순간 
객체에 메시지를 보내지 않고 값을 받아서 직접 처리 하려는 시도가 생길 것 같아요 ~

rima361 replied:
오! 맞아요. 외부로 값을 직접 노출하지 않게끔 수정했습니다~

jayden-17 commented on src/main/java/caculator/StringSupport.java:
```diff
@@ -0,0 +1,21 @@
+package caculator;
+
+public class StringSupport {
```
이 class 는 테스트에서만 사용하는 것 같아요 ~ 

테스트 코드쪽으로 옮겨서 유틸성 기능으로 사용해도 좋을 것 같아요 ~

jayden-17 commented on src/test/java/calculator/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,39 @@
+
+    @ParameterizedTest
+    @NullSource
+    @EmptySource
```
둘을 합쳐 놓은 `@NullAndEmptySource` 도 있습니다 😄

jayden-17 reviewed (approved):
확실히 2회차라 그런가 잘하시네요 ~ 
junit5 도 assertJ 도 익숙하게 사용하시고 👍 

전반적으로 잘해주셔서 merge 하도록 할게요 ~ 
별도로 개선해보시면 좋을 것 같아요 ~ 2단계도 잘 부탁드려요 ~ 

---
`Class 에 인격화를 하지 말라`는 것보다 `~er`을 사용하지 말라고 봤어요 ~ 
`무엇을 하는지`가 아니라 `무엇인지`에 대해 고민해서 이름을 지으라고 했었어요  ~

`StringSupport` 의 경우에는 너무 모호한 이름으로 생각이 되요, 실제로 구체적인 역할이 없어 보입니다. 
`StringSplitter` 는 정적 메소드만을 제공하기 때문에 객체의 이름이라기 보다는 namespace 로 생각됩니다.

그래서 두가지를 합쳐보면 어떨까요 ? 
`StringAddCalculatorInput` 이 무엇인지에 기반한 이름으로 생각되고, 이 객체가 하는 역할이 분명해질 것 같아요 ~

Commit:
refactor: package 이동

Commit:
refactor : package명 오타 수정

Commit:
refactor: positiveNumber의 value를 외부로 노출하지 않는 방식으로 리팩토링

Commit:
refactor: 코드리뷰 반영
- StringSupport 을 StringAddCalculatorInput 역할에 맞도록 rename
- StringSplitter 를 StringUtils 로 rename
